base_layer/wallet/src/ffi.rs

```diff
@@ -29,87 +29,402 @@
+        TransactionKernel,
+        TransactionOutput,
+        UnblindedOutput,
+    },
     types::{PrivateKey, PublicKey},
 };
+use tari_crypto::keys::SecretKey;
+use tari_p2p::initialization::CommsConfig;
+use tari_utilities::{hex::Hex, ByteArray};
+use tokio::runtime::Runtime;
 
-// use tari_core::transaction_protocol::sender::RawTransactionInfo;
+pub type TariWallet = Wallet;
+pub type WalletDateTime = NaiveDateTime;
 
-// TODO implement this
-pub struct WalletFfi {}
+/// -------------------------------- ByteVector ------------------------------------------------ ///
+pub struct ByteVector(Vec<c_uchar>); // declared like this so that it can be exposed to external header
 
-pub struct IdentityFfi {
-    public_key: PublicKey,
-    secret_key: PrivateKey,
+#[no_mangle]
+pub unsafe extern "C" fn bytevector_create() -> *mut ByteVector {
+    let bytes = ByteVector(Vec::new());
+    Box::into_raw(Box::new(bytes))
 }
 
-pub struct KeyManagerStateFfi {
-    master_seed: PrivateKey,
-    branch_seed: String,
-    index: c_uint,
+#[no_mangle]
+pub unsafe extern "C" fn bytevector_destroy(bytes: *mut ByteVector) {
+    if bytes.is_null() {
+        let b = Box::from_raw(bytes);
+    }
+}
+
+/// adds c_char to internal vec
+pub unsafe extern "C" fn bytevector_add_byte(bytes: *mut ByteVector, byte: c_uchar) {
+    (*bytes).0.push(byte);
+}
+
+/// returns c_char at position in internal vector
+#[no_mangle]
+pub unsafe extern "C" fn bytevector_get_at(ptr: *mut ByteVector, i: c_int) -> c_uchar {
+    (*ptr).0.clone()[i as usize]
+}
+
+/// Returns the number of items, zero-indexed
+#[no_mangle]
+pub unsafe extern "C" fn bytevector_get_length(vec: *const ByteVector) -> c_int {
+    if vec.is_null() {
+        return 0;
+    }
+    (&*vec).0.len() as c_int
+}
+
+/// -------------------------------------------------------------------------------------------- ///
+
+/// -------------------------------- Public Key ------------------------------------------------ ///
+pub type WalletPrivateKey = PublicKey;
+
+#[no_mangle]
+pub unsafe extern "C" fn public_key_create(bytes: *mut ByteVector) -> *mut WalletPublicKey {
+    let mut v = Vec::new();
+    if !bytes.is_null() {
+        v = (*bytes).0.clone();
+    }
+    let pk = WalletPublicKey::from_bytes(&v).unwrap();
+    Box::into_raw(Box::new(pk))
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn public_key_destroy(pk: *mut WalletPublicKey) {
+    if !pk.is_null() {
+        Box::from_raw(pk);
+    }
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn public_key_get_key(pk: *mut WalletPublicKey) -> *mut ByteVector {
+    let mut bytes = ByteVector(Vec::new());
+    if !pk.is_null() {
+        bytes.0 = (*pk).to_vec();
+    }
+    Box::into_raw(Box::new(bytes))
+}
+
+/// -------------------------------------------------------------------------------------------- ///
+
+/// -------------------------------- Private Key ----------------------------------------------- ///
+pub type WalletPublicKey = PrivateKey;
```

Comment:
Aliased a PrivateKey as a PublicKey. Again do this aliasing in the `use` command

Reply:
My bad. Fixed in next commit.